pallets/stateful-storage/src/types.rs

```diff
@@ -0,0 +1,288 @@
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::pallet_prelude::*;
+use scale_info::TypeInfo;
+use sp_core::{bounded::BoundedVec, Get};
+use sp_std::{cmp::*, collections::btree_map::BTreeMap, fmt::Debug, prelude::*};
+
+/// Defines the actions that can be applied to an Itemized storage
+#[derive(Clone, Encode, Decode, PartialEq, Debug, TypeInfo, Eq, PartialOrd, Ord)]
+pub enum ItemAction {
+	Add { data: Vec<u8> },
+	Remove { index: u16 },
+}
+
+/// This header is used to specify how long an item is inside the buffer and inserted into buffer
+/// before every item
+#[derive(Encode, Decode, PartialEq, MaxEncodedLen, Debug)]
+pub struct ItemHeader {
+	/// The length of this item, not including the size of this header.
+	payload_len: u16,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum ItemPageError {
+	ErrorParsing(&'static str),
+	InvalidAction(&'static str),
+	ArithmeticOverflow,
+}
+
+/// A page of items
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, Debug)]
+#[scale_info(skip_type_params(PageDataSize))]
+#[codec(mel_bound(PageDataSize: MaxEncodedLen))]
+pub struct ItemPage<PageDataSize: Get<u32>> {
+	data: BoundedVec<u8, PageDataSize>,
+}
+
+/// an internal struct which contains the parsed items in a page
+#[derive(Debug, PartialEq)]
+struct ParsedPage<'a> {
+	/// page current size
+	page_size: usize,
+	/// a map of item index to a slice of blob (header included)
+	items: BTreeMap<u16, &'a [u8]>,
+}
+
+impl<PageDataSize: Get<u32>> ItemPage<PageDataSize> {
+	/// creates new itemPage from BoundedVec
+	pub fn new(data: BoundedVec<u8, PageDataSize>) -> Self {
+		Self { data }
+	}
+
+	/// applies all actions to specified page and returns the updated page
+	pub fn apply_actions(
+		&self,
+		actions: &[ItemAction],
+	) -> Result<ItemPage<PageDataSize>, ItemPageError> {
+		let mut parsed = self.parse()?;
+
+		let mut updated_page_buffer = Vec::with_capacity(parsed.page_size);
+		let mut add_buffer = Vec::new();
+
+		for action in actions {
+			match action {
+				ItemAction::Remove { index } => {
+					ensure!(
+						parsed.items.contains_key(&index),
+						ItemPageError::InvalidAction("item index is invalid")
```

Comment:
I think these descriptive errors would be helpful for figuring out later in case for some users anything fails. open to changes if we'd like to replace the strings with anything else